Content.Shared/Containers/ItemSlot/ItemSlotsSystem.cs

```diff
@@ -245,27 +248,29 @@ private void Insert(EntityUid uid, ItemSlot slot, EntityUid item, EntityUid? use
         /// </summary>
         /// <remarks>
         ///     If a popup entity is given, and if the item slot is set to generate a popup message when it fails to
-        ///     pass the whitelist, then this will generate a popup.
+        ///     pass the whitelist or due to slot being locked, then this will generate an appropriate popup.
         /// </remarks>
         public bool CanInsert(EntityUid uid, EntityUid usedUid, EntityUid? user, ItemSlot slot, bool swap = false, EntityUid? popup = null)
         {
             if (slot.ContainerSlot == null)
                 return false;
 
-            if (slot.Locked)
-                return false;
-
-            if (!swap && slot.HasItem)
+            if ((slot.Whitelist != null && !slot.Whitelist.IsValid(usedUid)) ||
+                (slot.Blacklist != null && slot.Blacklist.IsValid(usedUid)))
+            {
```

Comment:
If you want to be less verbose you can do slot.Whitelist?.IsValid(usedUid) == false